libs/ml/src/ops/activations/logsoftmax.cpp

```diff
@@ -0,0 +1,119 @@
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2020 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "math/activation_functions/softmax.hpp"
+#include "math/standard_functions/log.hpp"
+#include "ml/ops/activations/logsoftmax.hpp"
+
+namespace fetch {
+namespace ml {
+namespace ops {
+
+template <typename TensorType>
+LogSoftmax<TensorType>::LogSoftmax(SizeType axis)
+  : axis_(axis)
+{}
+
+template <typename TensorType>
+LogSoftmax<TensorType>::LogSoftmax(SPType const &sp)
+  : Ops<TensorType>(sp)
+{
+  axis_ = sp.axis;
+}
+
+template <typename TensorType>
+std::shared_ptr<OpsSaveableParams> LogSoftmax<TensorType>::GetOpSaveableParams()
+{
+  auto sp_ptr  = std::make_shared<SPType>();
+  sp_ptr->axis = axis_;
+  return sp_ptr;
+}
+
+template <typename TensorType>
+std::shared_ptr<fetch::ml::ops::Ops<TensorType>> LogSoftmax<TensorType>::MakeSharedCopy(
+    std::shared_ptr<fetch::ml::ops::Ops<TensorType>> me)
+{
+  FETCH_UNUSED(me);
+  assert(me.get() == this);
+
+  auto copyshare = std::make_shared<MyType>(*this);  // calls default copy constructor of MyType
+
+  return copyshare;
+}
+
+template <typename TensorType>
+void LogSoftmax<TensorType>::Forward(VecTensorType const &inputs, TensorType &output)
+{
+  assert(output.shape() == ComputeOutputShape(inputs));
+  assert(inputs.size() == 1);
+  fetch::math::Softmax((*inputs.front()), output, axis_);
+  fetch::math::Log(output, output);
+}
+
+template <typename TensorType>
+std::vector<TensorType> LogSoftmax<TensorType>::Backward(VecTensorType const &inputs,
+                                                         TensorType const &   error_signal)
+{
+  assert(inputs.size() == 1);
+  assert(inputs.front()->shape() == error_signal.shape());
+
+  TensorType return_signal = error_signal.Copy();
+  TensorType t(error_signal.shape());
+  fetch::math::Softmax((*inputs.front()), t, axis_);
+
+  // return_signal.InlineMultiply(t);
```

Comment:
looks like an artifact, is it still needed?

Reply:
interesting, i guess not. will ditch it.